<a href="https://colab.research.google.com/github/abdalrahmenyousifMohamed/ML/blob/main/NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
# ! pip install kaggle

# ! mkdir ~/.kaggle


# ! cp kaggle.json ~/.kaggle/

# ! chmod 600 ~/.kaggle/kaggle.json

# ! kaggle datasets download -d jiashenliu/515k-hotel-reviews-data-in-europe

In [10]:
import pandas as pd
import numpy as np
import time

In [11]:
# ! unzip /content/515k-hotel-reviews-data-in-europe.zip

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv("/content/Hotel_Reviews.csv")
df.columns

Index(['Hotel_Address', 'Review_Date', 'Average_Score', 'Hotel_Name',
       'Reviewer_Nationality', 'Negative_Review', 'Total_Number_of_Reviews',
       'Positive_Review', 'Reviewer_Score', 'Tags', 'Leisure_trip', 'Couple',
       'Solo_traveler', 'Business_trip', 'Group', 'Family_with_young_children',
       'Family_with_older_children', 'With_a_pet'],
      dtype='object')

In [14]:
df.isnull().sum()

Hotel_Address                                    0
Additional_Number_of_Scoring                     0
Review_Date                                      0
Average_Score                                    0
Hotel_Name                                       0
Reviewer_Nationality                             0
Negative_Review                                  0
Review_Total_Negative_Word_Counts                0
Total_Number_of_Reviews                          0
Positive_Review                                  0
Review_Total_Positive_Word_Counts                0
Total_Number_of_Reviews_Reviewer_Has_Given       0
Reviewer_Score                                   0
Tags                                             0
days_since_review                                0
lat                                           3268
lng                                           3268
dtype: int64

In [15]:
#function with the idxmax() function to return the value with the highest count
#How many distinct values are there for the column Reviewer_Nationality and what are they?
df['Reviewer_Nationality'].value_counts().idxmax()

' United Kingdom '

In [16]:
most_freq_country_reviwers = df[df['Reviewer_Nationality']==' United Kingdom ']

In [17]:
most_freq_country_reviwers

Hotel_Address  \
3        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
6        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
7        s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
10       s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
12       s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
...                                                   ...   
515691  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515714  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515715  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515722  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515727  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   

        Additional_Number_of_Scoring Review_Date  Average_Score  \
3                                194   7/31/2017            7.7   
6                                194   7/17/2017            7.7   
7                                194   7/17/2017            7.7   
10                               194    7/7/2017            7.7   
12                               194    7/6/2017            7.7   
...                              ...         ...            ...   
515691                           168    7/5/2016            8.1   
515714                           168   4/14/2016            8.1   
515715                           168   3/30/2016            8.1   
515722                           168  11/26/2015            8.1   
515727                           168   9/27/2015            8.1   

                   Hotel_Name Reviewer_Nationality  \
3                 Hotel Arena      United Kingdom    
6                 Hotel Arena      United Kingdom    
7                 Hotel Arena      United Kingdom    
10                Hotel Arena      United Kingdom    
12                Hotel Arena      United Kingdom    
...                       ...                  ...   
515691  Atlantis Hotel Vienna      United Kingdom    
515714  Atlantis Hotel Vienna      United Kingdom    
515715  Atlantis Hotel Vienna      United Kingdom    
515722  Atlantis Hotel Vienna      United Kingdom    
515727  Atlantis Hotel Vienna      United Kingdom    

                                          Negative_Review  \
3        My room was dirty and I was afraid to walk ba...   
6        Cleaner did not change our sheet and duvet ev...   
7        Apart from the price for the brekfast Everyth...   
10                                     Nothing all great    
12       The floor in my room was filfy dirty Very bas...   
...                                                   ...   
515691                                        No Negative   
515714   Staff not always at reception and had to wait...   
515715    would have liked restaurant also serving dinner   
515722   noisy at night air causing loud banging in he...   
515727         Wifi Location No tea or coffee in the room   

        Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
3                                     210                     1403   
6                                      33                     1403   
7                                      11                     1403   
10                                      5                     1403   
12                                     28                     1403   
...                                   ...                      ...   
515691                                  0                     2823   
515714                                 16                     2823   
515715                                  8                     2823   
515722                                 19                     2823   
515727                                 10                     2823   

                                          Positive_Review  \
3        Great location in nice surroundings the bar a...   
6        The room is spacious and bright The hotel is ...   
7        Good location Set in a lovely park friendly s...   
10       Rooms were stunningly decorate

In [18]:
#What reviewer nationality is the most common in the dataset (print country and number of reviews)?
most_freq_country_reviwers.loc[:,['Total_Number_of_Reviews','Reviewer_Nationality']]

Total_Number_of_Reviews Reviewer_Nationality
3                          1403      United Kingdom 
6                          1403      United Kingdom 
7                          1403      United Kingdom 
10                         1403      United Kingdom 
12                         1403      United Kingdom 
...                         ...                  ...
515691                     2823      United Kingdom 
515714                     2823      United Kingdom 
515715                     2823      United Kingdom 
515722                     2823      United Kingdom 
515727                     2823      United Kingdom 

[245246 rows x 2 columns]

In [19]:
#What are the next top 10 most frequently found nationalities, and their frequency count?
top_most_10_na = df['Reviewer_Nationality'].value_counts()[:10].index

In [20]:
#What was the most frequently reviewed hotel for each of the top 10 most reviewer nationalities?
top_most_10_na = list(top_most_10_na)
df_top_10 = df[df.Reviewer_Nationality.isin(top_most_10_na)]

In [21]:
#What was the most frequently reviewed hotel for each of the top 10 most reviewer nationalities?
most_10_reviwed = list(df_top_10.Total_Number_of_Reviews.value_counts()[:10].index)
df.loc[most_10_reviwed,['Hotel_Name']].value_counts()[:1].index

MultiIndex([('Park Plaza County Hall London',)],
           names=['Hotel_Name'])

In [22]:
df.Hotel_Name.unique()

array(['Hotel Arena', 'K K Hotel George', 'Apex Temple Court Hotel', ...,
       'The Berkeley', 'Holiday Inn London Kensington',
       'Atlantis Hotel Vienna'], dtype=object)

In [23]:
len(df.Tags.unique())

55242

In [24]:
df[df.Tags.str.contains("Family with",na=False)].count()

Hotel_Address                                 87364
Additional_Number_of_Scoring                  87364
Review_Date                                   87364
Average_Score                                 87364
Hotel_Name                                    87364
Reviewer_Nationality                          87364
Negative_Review                               87364
Review_Total_Negative_Word_Counts             87364
Total_Number_of_Reviews                       87364
Positive_Review                               87364
Review_Total_Positive_Word_Counts             87364
Total_Number_of_Reviews_Reviewer_Has_Given    87364
Reviewer_Score                                87364
Tags                                          87364
days_since_review                             87364
lat                                           86799
lng                                           86799
dtype: int64

In [25]:
df.describe(include="all")

Hotel_Address  \
count                                              515738   
unique                                               1493   
top     163 Marsh Wall Docklands Tower Hamlets London ...   
freq                                                 4789   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

        Additional_Number_of_Scoring Review_Date  Average_Score  \
count                  515738.000000      515738  515738.000000   
unique                           NaN         731            NaN   
top                              NaN    8/2/2017            NaN   
freq                             NaN        2585            NaN   
mean                      498.081836         NaN       8.397487   
std                       500.538467         NaN       0.548048   
min                         1.000000         NaN       5.200000   
25%                       169.000000         NaN       8.100000   
50%                       341.000000         NaN       8.400000   
75%                       660.000000         NaN       8.800000   
max                      2682.000000         NaN       9.800000   

                                        Hotel_Name Reviewer_Nationality  \
count                                       515738               515738   
unique                                        1492                  227   
top     Britannia International Hotel Canary Wharf      United Kingdom    
freq                                          4789               245246   
mean                                           NaN                  NaN   
std                                            NaN                  NaN   
min                                            NaN                  NaN   
25%                                            NaN                  NaN   
50%                                            NaN                  NaN   
75%                                            NaN                  NaN   
max                                            NaN                  NaN   

       Negative_Review  Review_Total_Negative_Word_Counts  \
count           515738                      515738.000000   
unique          330011                                NaN   
top        No Negative                                NaN   
freq            127890                                NaN   
mean               NaN                          18.539450   
std                NaN                          29.690831   
min                NaN                           0.000000   
25%                NaN                           2.000000   
50%                NaN                           9.000000   
75%                NaN                          23.000000   
max                NaN                         408.000000   

        Total_Number_of_Reviews Positive_Review  \
count             515738.000000          515738   
unique                      NaN          412601   
top                         NaN     No Positive   
freq                        NaN           35946   
mean                2743.743944             NaN   
std                 2317.464868             NaN   
min                   43.000000             NaN   
25%                 1161.000000             NaN   
50%                 2134.000000             NaN   
75%                 3613.000000             NaN   
max                16670.000000             NaN   

        Review_Total_Positive_Word_Counts  \
count                       515738.000000   
unique                                NaN   
top                                   NaN   
freq                                  NaN   
mean                            17

In [26]:
sort_total_no_reviews=df.sort_values(by=['Total_Number_of_Reviews'], ascending=True)

In [27]:
df.head()

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
4   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring Review_Date  Average_Score   Hotel_Name  \
0                           194    8/3/2017            7.7  Hotel Arena   
1                           194    8/3/2017            7.7  Hotel Arena   
2                           194   7/31/2017            7.7  Hotel Arena   
3                           194   7/31/2017            7.7  Hotel Arena   
4                           194   7/24/2017            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to walk ba...   
4         New Zealand    You When I booked with your company on line y...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   
3                                210                     1403   
4                                140                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   
3   Great location in nice surroundings the bar a...   
4    Amazing location and building Romantic setting    

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   
3                                 26   
4                                  8   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   
3                                           1             3.8   
4                                           3             6.7   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   
3  [' Leisure trip ', ' Solo traveler ', ' Duplex...            3 days   
4  [' Leisure trip ', ' Couple ', ' Suite ', ' St...           10 days   

         lat       lng  
0  52.360576  4.915968  
1  52.360576  4.915968  
2  52.360576  4.915968  
3  52.360576  4.915968  
4  52.360576  4.915968

In [28]:
df[df["Hotel_Name"]=="Rydges Kensington Hotel"].describe()

Additional_Number_of_Scoring  Average_Score  \
count                         496.0   4.960000e+02   
mean                          205.0   8.100000e+00   
std                             0.0   1.778150e-15   
min                           205.0   8.100000e+00   
25%                           205.0   8.100000e+00   
50%                           205.0   8.100000e+00   
75%                           205.0   8.100000e+00   
max                           205.0   8.100000e+00   

       Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
count                         496.000000                    496.0   
mean                           21.104839                   1057.0   
std                            32.564173                      0.0   
min                             0.000000                   1057.0   
25%                             4.000000                   1057.0   
50%                            12.000000                   1057.0   
75%                            26.250000                   1057.0   
max                           384.000000                   1057.0   

       Review_Total_Positive_Word_Counts  \
count                         496.000000   
mean                           17.094758   
std                            19.606459   
min                             0.000000   
25%                             6.000000   
50%                            11.000000   
75%                            21.000000   
max                           150.000000   

       Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
count                                  496.000000      496.000000   
mean                                     5.360887        8.211895   
std                                      6.319187        1.617792   
min                                      1.000000        2.500000   
25%                                      1.000000        7.500000   
50%                                      3.000000        8.300000   
75%                                      7.000000        9.600000   
max                                     50.000000       10.000000   

                lat           lng  
count  4.960000e+02  4.960000e+02  
mean   5.149569e+01 -1.827878e-01  
std    7.112601e-15  2.778360e-17  
min    5.149569e+01 -1.827878e-01  
25%    5.149569e+01 -1.827878e-01  
50%    5.149569e+01 -1.827878e-01  
75%    5.149569e+01 -1.827878e-01  
max    5.149569e+01 -1.827878e-01

In [29]:
df[df["Hotel_Name"]=="Britannia International Hotel Canary Wharf"]

Hotel_Address  \
63942  163 Marsh Wall Docklands Tower Hamlets London ...   
63943  163 Marsh Wall Docklands Tower Hamlets London ...   
63944  163 Marsh Wall Docklands Tower Hamlets London ...   
63945  163 Marsh Wall Docklands Tower Hamlets London ...   
63946  163 Marsh Wall Docklands Tower Hamlets London ...   
...                                                  ...   
68726  163 Marsh Wall Docklands Tower Hamlets London ...   
68727  163 Marsh Wall Docklands Tower Hamlets London ...   
68728  163 Marsh Wall Docklands Tower Hamlets London ...   
68729  163 Marsh Wall Docklands Tower Hamlets London ...   
68730  163 Marsh Wall Docklands Tower Hamlets London ...   

       Additional_Number_of_Scoring Review_Date  Average_Score  \
63942                          2682    8/3/2017            7.1   
63943                          2682    8/3/2017            7.1   
63944                          2682    8/2/2017            7.1   
63945                          2682    8/2/2017            7.1   
63946                          2682    8/2/2017            7.1   
...                             ...         ...            ...   
68726                          2682    8/4/2015            7.1   
68727                          2682    8/4/2015            7.1   
68728                          2682    8/4/2015            7.1   
68729                          2682    8/4/2015            7.1   
68730                          2682    8/4/2015            7.1   

                                       Hotel_Name Reviewer_Nationality  \
63942  Britannia International Hotel Canary Wharf      United Kingdom    
63943  Britannia International Hotel Canary Wharf      United Kingdom    
63944  Britannia International Hotel Canary Wharf      United Kingdom    
63945  Britannia International Hotel Canary Wharf      United Kingdom    
63946  Britannia International Hotel Canary Wharf      United Kingdom    
...                                           ...                  ...   
68726  Britannia International Hotel Canary Wharf             Ireland    
68727  Britannia International Hotel Canary Wharf      United Kingdom    
68728  Britannia International Hotel Canary Wharf      United Kingdom    
68729  Britannia International Hotel Canary Wharf      United Kingdom    
68730  Britannia International Hotel Canary Wharf      United Kingdom    

                                         Negative_Review  \
63942   The car park was small and unpleasant People ...   
63943   We weren t told that the only spa facility op...   
63944   I asked how far the O2 was and got told a 7 m...   
63945   Hot stuffy room air con not working properly ...   
63946   Although the price seems like it is cheap you...   
...                                                  ...   
68726   Long wait for check in arrived at 6 30pm and ...   
68727                          concierge was uninformed    
68728                                I had no complaints   
68729   Really shabby and run down hotel Needs a tota...   
68730   Stains on the carpet peeling wallpaper scruff...   

       Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
63942                                 31                     9086   
63943                                 34                     9086   
63944                                 38                     9086   
63945                                 78                     9086   
63946                                 44                     9086   
...                                  ...                      ...   
68726                                 15                     9086   
68727                                  5                     9086   
68728                                  5                     9086   
68729                                 15                     9086   
68730                                 19                     9086   

                                         Positive_Review  \
63942   The location was excel



Print out the shape of the data frame you have just loaded (the shape is the number of rows and columns)

In [30]:
print("The shape of the data (rows , cols) is " + str(df.shape))

The shape of the data (rows , cols) is (515738, 17)


# calculate the frequency count for reviwer nationalities
1. How many distinct values are there for the column Reviewer_Nationality and what are they?

2. What reviewer nationality is the most common in the dataset (print country and no reviews)


In [31]:
nationality_freq = df["Reviewer_Nationality"].value_counts()
print("There are " + str(nationality_freq.size) + " different nationalities")
print(nationality_freq)

There are 227 different nationalities
 United Kingdom               245246
 United States of America      35437
 Australia                     21686
 Ireland                       14827
 United Arab Emirates          10235
                               ...  
 Cape Verde                        1
 Northern Mariana Islands          1
 Tuvalu                            1
 Guinea                            1
 Palau                             1
Name: Reviewer_Nationality, Length: 227, dtype: int64


what are the next top 10 most frequently found nationalities , and their frequency count?

In [32]:
print("The highest frequency reviewer nationality is " + str(nationality_freq.index[0]).strip() + " with " + str(nationality_freq[0]) + " reviews.")

print("The next 10 highest frequency reviewer nationalities are:")
print(nationality_freq[1:11].to_string())

The highest frequency reviewer nationality is United Kingdom with 245246 reviews.
The next 10 highest frequency reviewer nationalities are:
 United States of America     35437
 Australia                    21686
 Ireland                      14827
 United Arab Emirates         10235
 Saudi Arabia                  8951
 Netherlands                   8772
 Switzerland                   8678
 Germany                       7941
 Canada                        7894
 France                        7296


what are the most frequently reviewed hotel for each of the top 10 most reviewer nationalities?

In [33]:
# Normally with pandas you will avoid an explicit loop, but wanted to show creating a new dataframe using criteria (don't do this with large amounts of data because it could be very slow)
for nat in nationality_freq[:10].index:
    # First , extract all rows that matched the criteria into a new df
    nat_df = df[df['Reviewer_Nationality']==nat]
    # Now get the hotel freq
    hotel_freq = nat_df["Hotel_Name"].value_counts()
    
    print("The most reviewed hotel for " + str(nat).strip() + " was " + str(hotel_freq.index[0]) + " with " + str(hotel_freq[0]) + " reviews.")
    

The most reviewed hotel for United Kingdom was Britannia International Hotel Canary Wharf with 3833 reviews.
The most reviewed hotel for United States of America was Hotel Esther a with 423 reviews.
The most reviewed hotel for Australia was Park Plaza Westminster Bridge London with 167 reviews.
The most reviewed hotel for Ireland was Copthorne Tara Hotel London Kensington with 239 reviews.
The most reviewed hotel for United Arab Emirates was Millennium Hotel London Knightsbridge with 129 reviews.
The most reviewed hotel for Saudi Arabia was The Cumberland A Guoman Hotel with 142 reviews.
The most reviewed hotel for Netherlands was Jaz Amsterdam with 97 reviews.
The most reviewed hotel for Switzerland was Hotel Da Vinci with 97 reviews.
The most reviewed hotel for Germany was Hotel Da Vinci with 86 reviews.
The most reviewed hotel for Canada was St James Court A Taj Hotel London with 61 reviews.


How many reviews are there per hotel (frequency count of hotel) in the dataset?

In [34]:
df.columns

Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Review_Date',
       'Average_Score', 'Hotel_Name', 'Reviewer_Nationality',
       'Negative_Review', 'Review_Total_Negative_Word_Counts',
       'Total_Number_of_Reviews', 'Positive_Review',
       'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags',
       'days_since_review', 'lat', 'lng'],
      dtype='object')

In [35]:
df[df['Hotel_Name'] =='K K Hotel George'].count()

Hotel_Address                                 566
Additional_Number_of_Scoring                  566
Review_Date                                   566
Average_Score                                 566
Hotel_Name                                    566
Reviewer_Nationality                          566
Negative_Review                               566
Review_Total_Negative_Word_Counts             566
Total_Number_of_Reviews                       566
Positive_Review                               566
Review_Total_Positive_Word_Counts             566
Total_Number_of_Reviews_Reviewer_Has_Given    566
Reviewer_Score                                566
Tags                                          566
days_since_review                             566
lat                                           566
lng                                           566
dtype: int64

In [36]:
#First create a new dataframe based on the old one , removing the uneeded columns
hotel_freq_df = df.drop(["Hotel_Address", "Additional_Number_of_Scoring", "Review_Date", "Average_Score", "Reviewer_Nationality", "Negative_Review", "Review_Total_Negative_Word_Counts", "Positive_Review", "Review_Total_Positive_Word_Counts", "Total_Number_of_Reviews_Reviewer_Has_Given", "Reviewer_Score", "Tags", "days_since_review", "lat", "lng"], axis = 1)
print(hotel_freq_df)
#print(hotel_freq_df.groupby('Hotel_Name').transform('count'))
# Group the rows by Hotel_Name , count them and put the result in a new column Total_Reviews_Found
hotel_freq_df['Total_Reviews_Found'] = hotel_freq_df.groupby('Hotel_Name').transform('count')

# Get rid of all the duplicated rows
hotel_freq_df = hotel_freq_df.drop_duplicates(subset = ["Hotel_Name"])
#display(hotel_freq_df.sort_values(by=['Total_Number_of_Reviews'],ascending=False))

                   Hotel_Name  Total_Number_of_Reviews
0                 Hotel Arena                     1403
1                 Hotel Arena                     1403
2                 Hotel Arena                     1403
3                 Hotel Arena                     1403
4                 Hotel Arena                     1403
...                       ...                      ...
515733  Atlantis Hotel Vienna                     2823
515734  Atlantis Hotel Vienna                     2823
515735  Atlantis Hotel Vienna                     2823
515736  Atlantis Hotel Vienna                     2823
515737  Atlantis Hotel Vienna                     2823

[515738 rows x 2 columns]


In [37]:
hotel_freq_df.groupby('Hotel_Name').transform('count')

Total_Number_of_Reviews  Total_Reviews_Found
0                             1                    1
405                           1                    1
971                           1                    1
2008                          1                    1
3778                          1                    1
...                         ...                  ...
511962                        1                    1
512401                        1                    1
512545                        1                    1
512645                        1                    1
515413                        1                    1

[1492 rows x 2 columns]

while there is Average_Score column for each hotel in the dataset , you can also calculate an average score ( getting the average of all reviewer scores in the dataset for each hotel ) . 
Add a new column to your dataframe with the column header Calc_AVerage_Score that contains that calculated average.
print out the columns Hoel_Name , Average_Score , and Calc_Average_Score.

In [38]:
# define a function that takes a row and performs some calculations with it
def get_difference_review_avg(row):
    return row["Average_Score"] - row['Calc_Average_Score']

# 'mean' is  mathematical word for 'average'
df['Calc_Average_Score'] = round(df.groupby('Hotel_Name').Reviewer_Score.transform('mean'),1)

# Add a new column with the difference between the two average scores
df["Average_Score_Difference"] = df.apply(get_difference_review_avg,axis=1)
# create a df without all the duplicates of Hotel_Name (so only 1 row per hotel)
review_scores_df = df.drop_duplicates(subset=["Hotel_Name"])
# Sort the dataframe to find the lowest and highest average score difference
review_scores_df = review_scores_df.sort_values(by=["Average_Score_Difference"])
display(review_scores_df[["Average_Score_Difference" , "Average_Score","Calc_Average_Score","Hotel_Name"]])

Average_Score_Difference  Average_Score  Calc_Average_Score  \
495945                      -0.8            7.7                 8.5   
111027                      -0.7            8.8                 9.5   
43688                       -0.7            7.5                 8.2   
178253                      -0.7            7.9                 8.6   
218258                      -0.5            7.0                 7.5   
...                          ...            ...                 ...   
151416                       0.7            7.8                 7.1   
22189                        0.8            7.1                 6.3   
250308                       0.9            8.6                 7.7   
68936                        0.9            6.8                 5.9   
3813                         1.3            7.2                 5.9   

                                               Hotel_Name  
495945                         Best Western Hotel Astoria  
111027  Hotel Stendhal Place Vend me Paris MGallery by...  
43688                       Mercure Paris Porte d Orleans  
178253                    Renaissance Paris Vendome Hotel  
218258                                Hotel Royal Elys es  
...                                                   ...  
151416                        Best Western Allegro Nation  
22189              Holiday Inn Paris Montparnasse Pasteur  
250308          MARQUIS Faubourg St Honor Relais Ch teaux  
68936                                       Villa Eugenie  
3813                                   Kube Hotel Ice Bar  

[1492 rows x 4 columns]

With only 1 hotel having a difference of score greater than 1, it means we can probably ignore the difference and use the calculated average score.



1. Calculate and print out how many rows have column Negative_Review values of "No Negative"

2. calculate and print out how many rows have Positive_Review values of "No Positive"

3. Calculate and print out how many rows have column Positive_Review values of "No Positive" and Negative_Review values of "No Negative"

In [39]:
len(df[df['Negative_Review']=="No Negative"])

127890

In [40]:
import time
start = time.time()
no_negative_reviews = df.apply(lambda x: True if x["Negative_Review"]=="No Negative" else False , axis=1)
print(" Number of No Negative reviews: " + str(len(no_negative_reviews[no_negative_reviews==True].index)))

no_positive_reviews = df.apply(lambda x : True if x["Positive_Review"]=="No Positive" else False , axis=1)
print("Number Of No Positive reviews: " + str(len(no_positive_reviews[no_negative_reviews==True].index)))

both_no_reviews = df.apply(lambda x: True if x['Negative_Review']=="No Negative" and x["Positive_Review"]=="No Positive" else False,axis=1)
print("Number of both No negative_Review and No positive_Review : " + str(len(both_no_reviews[both_no_reviews==True].index)))

end = time.time()

print(" Lambdas took " + str(round(end - start , 2)) + " Seconds ")

 Number of No Negative reviews: 127890
Number Of No Positive reviews: 127890
Number of both No negative_Review and No positive_Review : 127
 Lambdas took 16.37 Seconds 


**Another way**

Another way count items without Lambdas, and use sum to count the rows:

In [41]:
# without lambdas (using a mixture of notations to show you can use both)
start = time.time()
no_negative_reviews = sum(df.Negative_Review == "No Negative")
print("Number of No Negative reviews: " + str(no_negative_reviews))

no_positive_reviews = sum(df["Positive_Review"] == "No Positive")
print("Number of No Positive reviews: " + str(no_positive_reviews))

both_no_reviews = sum((df.Negative_Review == "No Negative") & (df.Positive_Review == "No Positive"))
print("Number of both No Negative and No Positive reviews: " + str(both_no_reviews))

end = time.time()
print("Sum took " + str(round(end - start, 2)) + " seconds")

Number of No Negative reviews: 127890
Number of No Positive reviews: 35946
Number of both No Negative and No Positive reviews: 127
Sum took 0.37 seconds


**You may have noticed that there are 127 rows that have both "No Negative" and "No Positive" values for the columns Negative_Review and Positive_Review respectively. That means that the reviewer gave the hotel a numerical score, but declined to write either a positive or negative review. Luckily this is a small amount of rows (127 out of 515738, or 0.02%), so it probably won't skew our model or results in any particular direction, but you might not have expected a data set of reviews to have rows with no reviews, so it's worth exploring the data to discover rows like this.**


# Filtering & Sentiment Analysis Operations

As you've probably noticed, the dataset has a few issues. Some columns are filled with useless information, others seem incorrect. If they are correct, it's unclear how they were calculated, and answers cannot be independently verified by your own calculations.

# Exercise: a bit more data processing
clean the data just a bit more , Add columns that will be useful later , change the values in other columns , and drop certain columns completely

1. initial column processing
i. Drop "lat" and "lng"
ii. Replace "Hotel_Address" values with the following values ( if the address contains the same of the city and the country , change it to just the city and the country).

. These are the only cities and countries in thr dataset:
- Amsterdam , Nertherlands
- London , United Kingdom
- Vienna, Austria
- Paris, France 
- Milan , Italy
- Barcelona, Spain


2. process Hotel Meta-review columns
3. Drop "Additional_Number_of_Reviews"
4. Replace "Total_Number_of_Reviews" with the total number of reviews for that hotel that are actually in the dataset
5. Replace Average_Score with our own calculated score

In [42]:
def replace_address(row):
    if "Netherlands" in row["Hotel_Address"]:
        return "Amsterdam, Netherlands"
    elif "Barcelona" in row["Hotel_Address"]:
        return "Barcelona, Spain"
    elif "United Kingdom" in row["Hotel_Address"]:
        return "London, United Kingdom"
    elif "Milan" in row["Hotel_Address"]:        
        return "Milan, Italy"
    elif "France" in row["Hotel_Address"]:
        return "Paris, France"
    elif "Vienna" in row["Hotel_Address"]:
        return "Vienna, Austria" 
    else:
        return row.Hotel_Address


# Load the hotel reviews from CSV
start = time.time()
df = pd.read_csv('/content/Hotel_Reviews.csv')
# dropping columns we will not use:
df.drop(["lat", "lng"], axis = 1, inplace=True)
# Replace all the addresses with a shortened, more useful form
df["Hotel_Address"] = df.apply(replace_address, axis = 1)
# Drop `Additional_Number_of_Scoring`
df.drop(["Additional_Number_of_Scoring"], axis = 1, inplace=True)
# Replace `Total_Number_of_Reviews` and `Average_Score` with our own calculated values
hotel_freq_df = df.drop(["Hotel_Address", "Review_Date", "Average_Score", "Reviewer_Nationality", "Negative_Review", "Review_Total_Negative_Word_Counts", "Positive_Review", "Review_Total_Positive_Word_Counts", "Total_Number_of_Reviews_Reviewer_Has_Given", "Reviewer_Score", "Tags", "days_since_review"], axis = 1)

df.Total_Number_of_Reviews = hotel_freq_df.groupby('Hotel_Name').transform('count')
df.Average_Score = round(df.groupby('Hotel_Name').Reviewer_Score.transform('mean'), 1)

In [43]:
df.head()

Hotel_Address Review_Date  Average_Score   Hotel_Name  \
0  Amsterdam, Netherlands    8/3/2017            7.8  Hotel Arena   
1  Amsterdam, Netherlands    8/3/2017            7.8  Hotel Arena   
2  Amsterdam, Netherlands   7/31/2017            7.8  Hotel Arena   
3  Amsterdam, Netherlands   7/31/2017            7.8  Hotel Arena   
4  Amsterdam, Netherlands   7/24/2017            7.8  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to walk ba...   
4         New Zealand    You When I booked with your company on line y...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                      405   
1                                  0                      405   
2                                 42                      405   
3                                210                      405   
4                                140                      405   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   
3   Great location in nice surroundings the bar a...   
4    Amazing location and building Romantic setting    

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   
3                                 26   
4                                  8   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   
3                                           1             3.8   
4                                           3             6.7   

                                                Tags days_since_review  
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days  
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days  
2  [' Leisure trip ', ' Family with young childre...            3 days  
3  [' Leisure trip ', ' Solo traveler ', ' Duplex...            3 days  
4  [' Leisure trip ', ' Couple ', ' Suite ', ' St...           10 days

In [44]:
# Now you can query country level data:
display(df.groupby("Hotel_Address").agg({"Hotel_Name":"nunique"}))

Hotel_Name
Hotel_Address                     
Amsterdam, Netherlands         105
Barcelona, Spain               211
London, United Kingdom         400
Milan, Italy                   162
Paris, France                  458
Vienna, Austria                158

# Process review columns
Drop Review_Total_Negative_Word_Counts, Review_Total_Positive_Word_Counts, Review_Date and days_since_review

Keep Reviewer_Score, Negative_Review, and Positive_Review as they are,

Keep Tags for now

We'll be doing some additional filtering operations on the tags in the next section and then tags will be dropped
Process reviewer columns

Drop Total_Number_of_Reviews_Reviewer_Has_Given

Keep Reviewer_Nationality

In [45]:
df.Tags.unique()

array(["[' Leisure trip ', ' Couple ', ' Duplex Double Room ', ' Stayed 6 nights ']",
       "[' Leisure trip ', ' Couple ', ' Duplex Double Room ', ' Stayed 4 nights ']",
       "[' Leisure trip ', ' Family with young children ', ' Duplex Double Room ', ' Stayed 3 nights ', ' Submitted from a mobile device ']",
       ...,
       "[' Leisure trip ', ' Solo traveler ', ' Superior Double Room with Garden View ', ' Stayed 1 night ', ' Submitted from a mobile device ']",
       "[' Business trip ', ' Solo traveler ', ' Superior Double Room with Garden View ', ' Stayed 2 nights ']",
       "[' Leisure trip ', ' Family with young children ', ' Standard Double or Twin Room ', ' Stayed 10 nights ']"],
      dtype=object)

# Filtering tags

Remember that the goal of the dataset is to add sentiment and columns that will help you choose the best hotel (**for yourself or maybe a client tasking you to make a hotel recommendation bot**). You need to ask yourself if the tags are useful or not in the final dataset. Here is one interpretation (if you needed the dataset for other reasons different tags might stay in/out of the selection):


1. The type of trip is relevant, and that should stay
2. The type of guest group is important, and that should stay
3. The type of room, suite, or studio that the guest stayed in is irrelevant (all hotels have basically the same rooms)
3. The device the review was submitted on is irrelevant
4. The number of nights reviewer stayed for could be relevant if you attributed longer stays with them liking the hotel more, but it's a stretch, and probably irrelevant
**In summary, keep 2 kinds of tags and remove the others.**

First, you don't want to count the tags until they are in a better format, so that means removing the square brackets and quotes. You can do this several ways, but you want the fastest as it could take a long time to process a lot of data. Luckily, pandas has an easy way to do each of these steps.



In [49]:
df.Tags.str.strip("[]")[0]

"' Leisure trip ', ' Couple ', ' Duplex Double Room ', ' Stayed 6 nights '"

In [50]:
df.Tags.str.strip("[']")[0]

" Leisure trip ', ' Couple ', ' Duplex Double Room ', ' Stayed 6 nights "

In [51]:
df.Tags.str.replace(" ', '", ",", regex = False)

0         [' Leisure trip, Couple, Duplex Double Room, S...
1         [' Leisure trip, Couple, Duplex Double Room, S...
2         [' Leisure trip, Family with young children, D...
3         [' Leisure trip, Solo traveler, Duplex Double ...
4         [' Leisure trip, Couple, Suite, Stayed 2 night...
                                ...                        
515733    [' Leisure trip, Family with older children, 2...
515734    [' Leisure trip, Family with young children, S...
515735    [' Leisure trip, Family with older children, 2...
515736    [' Leisure trip, Group, Standard Triple Room, ...
515737    [' Leisure trip, Family with young children, 2...
Name: Tags, Length: 515738, dtype: object

In [52]:
# Remove opening and closing brackets
df.Tags = df.Tags.str.strip("[']")
# remove all quotes too
df.Tags = df.Tags.str.replace(" ', '", ",", regex = False)

In [54]:
df.Tags[0]

' Leisure trip, Couple, Duplex Double Room, Stayed 6 nights '

In [56]:
# process the tags into new columns
# Leisure trip, Couple, Solo traveler, Business trip, Group combined with Travelers with friends, 
# Family with young children, Family with older children, With a pet
df["Leisure_trip"] = df.Tags.apply(lambda tag: 1 if "Leisure trip" in tag else 0)
df["Couple"] = df.Tags.apply(lambda tag: 1 if "Couple" in tag else 0)
df["Solo_traveler"] = df.Tags.apply(lambda tag: 1 if "Solo traveler" in tag else 0)
df["Business_trip"] = df.Tags.apply(lambda tag: 1 if "Business trip" in tag else 0)
df["Group"] = df.Tags.apply(lambda tag: 1 if "Group" in tag or "Travel with friends" in tag else 0)
df["Family_with_young_children"] = df.Tags.apply(lambda tag: 1 if "Family with young children" in tag else 0)
df["Family_with_older_children"] = df.Tags.apply(lambda tag: 1 if "Family with older children" in tag else 0)
df["With_a_pet"] = df.Tags.apply(lambda tag: 1 if "With a pet" in tag else 0)


In [64]:
#df.drop(["Review_Total_Negative_Word_Counts", "Review_Total_Positive_Word_Counts", "days_since_review", "Total_Number_of_Reviews_Reviewer_Has_Given"], axis = 1, inplace=True)
df.drop(["Review_Date"], axis = 1, inplace=True)

# Saving new data file with calculated columns
print("Saving results to Hotel_Reviews_Filtered.csv")
df.to_csv(r'/content/Hotel_Reviews_Filtered.csv', index = False)

Saving results to Hotel_Reviews_Filtered.csv


# Sentiment Analysis Operations

- in this final section , you will apply sentiment analysis to the review columns and save the results in dataset

# Exercise: load and save the filtered data

- Note that now you are loading the filtered dataset that was saved in previous section , not the orginal dataset.

In [7]:
import time
import pandas as pd
import nltk as nltk
from nltk.corpus import stopwords
from nltk.sentiment.vader import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')

#Load the filtered hotel reviews from CSV
df2= pd.read_csv("/content/drive/MyDrive/Hotel_Reviews_Filtered.csv")

# You code will be added here

# Finally remeber to save the hotel reviews with new NLP data added
# print("Saving results to Hotel_Reviews_NLP.csv")
# df2.to_csv(r'Hotel_Reviews_NLP.csv' , index = False)

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


# Removing stop words
 Removing stop words,or common English words that do not change the sentiment of a sentence, is the first step. By removing them, the sentiment analysis should run faster, but not be less accurate (as the stop words do not affect sentiment, but they do slow down the analysis).

 **Stop words are available in abundance in any human language. By removing these words, we remove the low-level information from our text in order to give more focus to the important information.Feb 10, 2021**


In [10]:
# remove stop words - can be slow for a lot of text!
nltk.download('stopwords')
start = time.time()
cache = set(stopwords.words("english"))
def remove_stopwords(review):
  text = " ".join([word for word in review.split() if word not in cache])
  return text

# remove the stop words from both columns
df2.Negative_Review = df2.Negative_Review.apply(remove_stopwords)
df2.Positive_Review = df2.Positive_Review.apply(remove_stopwords)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [11]:
df2.head()

Hotel_Address Review_Date  Average_Score   Hotel_Name  \
0  Amsterdam, Netherlands    8/3/2017            7.8  Hotel Arena   
1  Amsterdam, Netherlands    8/3/2017            7.8  Hotel Arena   
2  Amsterdam, Netherlands   7/31/2017            7.8  Hotel Arena   
3  Amsterdam, Netherlands   7/31/2017            7.8  Hotel Arena   
4  Amsterdam, Netherlands   7/24/2017            7.8  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia   I angry made post available via possible sites...   
1             Ireland                                         No Negative   
2           Australia   Rooms nice elderly bit difficult rooms two sto...   
3      United Kingdom   My room dirty I afraid walk barefoot floor loo...   
4         New Zealand   You When I booked company line showed pictures...   

   Total_Number_of_Reviews                                    Positive_Review  \
0                      405                  Only park outside hotel beautiful   
1                      405  No real complaints hotel great great location ...   
2                      405  Location good staff ok It cute hotel breakfast...   
3                      405  Great location nice surroundings bar restauran...   
4                      405         Amazing location building Romantic setting   

   Reviewer_Score                                               Tags  \
0             2.9   Leisure trip, Couple, Duplex Double Room, Sta...   
1             7.5   Leisure trip, Couple, Duplex Double Room, Sta...   
2             7.1   Leisure trip, Family with young children, Dup...   
3             3.8   Leisure trip, Solo traveler, Duplex Double Ro...   
4             6.7   Leisure trip, Couple, Suite, Stayed 2 nights,...   

   Leisure_trip  Couple  Solo_traveler  Business_trip  Group  \
0             1       1              0              0      0   
1             1       1              0              0      0   
2             1       0              0              0      0   
3             1       0              1              0      0   
4             1       1              0              0      0   

   Family_with_young_children  Family_with_older_children  With_a_pet  
0                           0                           0           0  
1                           0                           0           0  
2                           1                           0           0  
3                           0                           0           0  
4                           0                           0           0

In [12]:
# Create the vader sentiment analyser ( there are others in NLTK you can try too)

vader_sentiment = SentimentIntensityAnalyzer()

# There are 3 possibilities of input for a review:
# It could be "No Negative", in which case, return 0 
# It could be "No Positive", in which case, return 0
# It could be a review, in which case calculate the sentiment
def calc_sentiment(review):
  if review == "No Positive" or review == "No Negative":
    return 0
  return vader_sentiment.polarity_scores(review)["compound"]

In [13]:
# Add a negative sentiment and positive sentiment column

print("Calculating sentiment columns for both positive and negative reviews"
)

start = time.time()
df2["Negative_Sentiment"] = df2.Negative_Review.apply(calc_sentiment)
df2["Positive_Sentiment"] = df2.Positive_Review.apply(calc_sentiment)
end = time.time()
print("Calculating sentiment took " + str(round(end -start,2)) + " Seconds" )

Calculating sentiment columns for both positive and negative reviews


In [16]:
df2 = df2.sort_values(by=["Negative_Sentiment"], ascending=True)
print(df2[["Negative_Review", "Negative_Sentiment"]])
df2 = df2.sort_values(by=["Positive_Sentiment"], ascending=True)
print(df2[["Positive_Review", "Positive_Sentiment"]])

                                          Negative_Review  Negative_Sentiment
186584  So bad experience memories I hotel The first n...             -0.9920
129503  First charged twice room booked booking second...             -0.9896
307286  The staff Had bad experience even booking Janu...             -0.9889
201953  Everything DO NOT STAY AT THIS HOTEL I never i...             -0.9886
452092  No WLAN room Incredibly rude restaurant staff ...             -0.9884
...                                                   ...                 ...
138365  Wifi terribly slow I speed test network upload...              0.9938
79215   I find anything hotel first I walked past hote...              0.9938
278506  The property great location There bakery next ...              0.9945
339189  Guys I like hotel I wish return next year Howe...              0.9948
480509  I travel lot far visited countless number hote...              0.9957

[515738 rows x 2 columns]
                                     

In [17]:
# Reorder the columns (This is cosmetic, but to make it easier to explore the data later)
df2 = df2.reindex(["Hotel_Name", "Hotel_Address", "Total_Number_of_Reviews", "Average_Score", "Reviewer_Score", "Negative_Sentiment", "Positive_Sentiment", "Reviewer_Nationality", "Leisure_trip", "Couple", "Solo_traveler", "Business_trip", "Group", "Family_with_young_children", "Family_with_older_children", "With_a_pet", "Negative_Review", "Positive_Review"], axis=1)

print("Saving results to Hotel_Reviews_NLP.csv")
df2.to_csv(r"/Hotel_Reviews_NLP.csv", index = False)

Saving results to Hotel_Reviews_NLP.csv
